File: tests/rosetta/transpiler/Java/cramers-rule.java
public class Main {
    static double[][] m = new double[][]{new double[]{2.0, -1.0, 5.0, 1.0}, new double[]{3.0, 2.0, 2.0, -6.0}, new double[]{1.0, 3.0, 3.0, -1.0}, new double[]{5.0, -2.0, -3.0, 3.0}};
    static double[] v = new double[]{-3.0, -32.0, -47.0, 49.0};
    static double d = det(m);
    static double[] x = new double[]{};
    static int i = 0;
    static String s = "[";
    static int j = 0;

    static double det(double[][] m) {
        int n = m.length;
        if (n == 1) {
            return m[0][0];
        }
        double total = 0.0;
        double sign = 1.0;
        int c = 0;
        while (c < n) {
            double[][] sub = new double[][]{};
            int r = 1;
            while (r < n) {
                double[] row = new double[]{};
                int cc = 0;
                while (cc < n) {
                    if (cc != c) {
                        row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(m[r][cc])).toArray();
                    }
                    cc = cc + 1;
                }
                sub = appendObj(sub, row);
                r = r + 1;
            }
            total = total + sign * m[0][c] * det(sub);
            sign = sign * (-1.0);
            c = c + 1;
        }
        return total;
    }

    static double[][] replaceCol(double[][] m, int col, double[] v) {
        double[][] res = new double[][]{};
        int r = 0;
        while (r < m.length) {
            double[] row = new double[]{};
            int c = 0;
            while (c < m[r].length) {
                if (c == col) {
                    row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(v[r])).toArray();
                } else {
                    row = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(row), java.util.stream.DoubleStream.of(m[r][c])).toArray();
                }
                c = c + 1;
            }
            res = appendObj(res, row);
            r = r + 1;
        }
        return res;
    }
    public static void main(String[] args) {
        while (i < v.length) {
            double[][] mc = replaceCol(m, i, v);
            x = java.util.stream.DoubleStream.concat(java.util.Arrays.stream(x), java.util.stream.DoubleStream.of(det(mc) / d)).toArray();
            i = i + 1;
        }
        while (j < x.length) {
            s = s + String.valueOf(x[j]);
            if (j < x.length - 1) {
                s = s + " ";
            }
            j = j + 1;
        }
        s = s + "]";
        System.out.println(s);
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }
}

File: tests/rosetta/transpiler/FS/cramers-rule.fs
// Generated 2025-07-28 11:14 +0700

exception Return

let rec det (m: float array array) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable m = m
    try
        let n: int = Seq.length m
        if n = 1 then
            __ret <- (m.[0]).[0]
            raise Return
        let mutable total: float = 0.0
        let mutable sign: float = 1.0
        let mutable c: int = 0
        while c < n do
            let mutable sub: float array array = [||]
            let mutable r: int = 1
            while r < n do
                let mutable row: float array = [||]
                let mutable cc: int = 0
                while cc < n do
                    if cc <> c then
                        row <- Array.append row [|(m.[r]).[cc]|]
                    cc <- cc + 1
                sub <- Array.append sub [|row|]
                r <- r + 1
            total <- total + (float ((sign * ((m.[0]).[c])) * (float (det sub))))
            sign <- sign * (-1.0)
            c <- c + 1
        __ret <- total
        raise Return
        __ret
    with
        | Return -> __ret
and replaceCol (m: float array array) (col: int) (v: float array) =
    let mutable __ret : float array array = Unchecked.defaultof<float array array>
    let mutable m = m
    let mutable col = col
    let mutable v = v
    try
        let mutable res: float array array = [||]
        let mutable r: int = 0
        while r < (Seq.length m) do
            let mutable row: float array = [||]
            let mutable c: int = 0
            while c < (Seq.length (m.[r])) do
                if c = col then
                    row <- Array.append row [|v.[r]|]
                else
                    row <- Array.append row [|(m.[r]).[c]|]
                c <- c + 1
            res <- Array.append res [|row|]
            r <- r + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let m: float array array = [|[|2.0; -1.0; 5.0; 1.0|]; [|3.0; 2.0; 2.0; -6.0|]; [|1.0; 3.0; 3.0; -1.0|]; [|5.0; -2.0; -3.0; 3.0|]|]
let v: float array = [|-3.0; -32.0; -47.0; 49.0|]
let d: float = det m
let mutable x: float array = [||]
let mutable i: int = 0
while i < (Seq.length v) do
    let mc: float array array = replaceCol m i v
    x <- Array.append x [|unbox<float> ((float (det mc)) / d)|]
    i <- i + 1
let mutable s: string = "["
let mutable j: int = 0
while j < (Seq.length x) do
    s <- s + (string (x.[j]))
    if j < ((Seq.length x) - 1) then
        s <- s + " "
    j <- j + 1
s <- s + "]"
printfn "%s" s
